### ISSUE TITLE
"Explicit" mappings contain all default values after upgrading to 5.x

### ISSUE THREAD

**Issue by AlonCG** [AutoMapper/AutoMapper#1570]
Not sure where I should have posted this (here or SO, so let me know where else I could have looked) ... any direction pointing, RTFMing, or suggestions would be greatly appreciated.

I recently updated to 5.02 from 4.1.1 and bit the bullet to attempt DI. Thinking I had to drop all the "static" stuff (guess based on old information) I did that as well ... but wrongly somewhere. 

I have a dozen or so "Profiles" that "explicitly" map an entity to our model structure. They follow a contrived pattern for our models that looks similar to ... DetailModel : Model : BaseModel : IdModel (":" for inherits). So the Profile's "CreateMap"s used "ForAllMembers", "IncludeBase", and "ForMember" to construct our mappings. 

Any mapping using these profiles ONLY returns the Model with it's default values (empty stings, default guids, 0, etc). None of the set entity properties are actually updated. This worked fine in 4.1.1.

I have assumptions, but they have turned out to be wrong, so I'll just post some code snippets, and maybe you can see where I am messing things up!

#### nInject Binding ####
```
var profileType = typeof(Profile);
var profiles = from t in Assembly.GetExecutingAssembly().GetTypes()
	where t.IsClass && profileType.IsAssignableFrom(t) && t != profileType
	select Activator.CreateInstance(t) as Profile;
var config = new MapperConfiguration( cfg => {
	cfg.DisableConstructorMapping(); // does not seem to change anything 
	foreach (var profile in profiles) {
		cfg.AddProfile(profile);
	}
});
config.AssertConfigurationIsValid();
var mapper = config.CreateMapper(); 
kernel.Bind<IMapper>().ToConstant(mapper);
```
#### Working CreateMap (contained within a profile) ####
```
CreateMap<Entity, EntityModel>()
	.ConstructUsing(source => new DataFilterModel {
		Id = source.EntityModelId,
		ParentId = source.ParentId,
		DisplayName = source.DisplayName,
		ColumnName = source.ColumnName,
		DataType = source.DataType
	})
	.ForAllMembers(opt => opt.Ignore());
```
#### Mappings that no longer Work all have the same pattern as below ####
```
public class MyMappingProfile : Profile 
{
	public MyMappingProfile()
	{
		var idMap = CreateMap<Entity, IdModel>();
		idMap.ForAllMembers(opt => opt.Ignore());
		idMap.ForMember( target => target.Id, m => m.MapFrom( source => source.EntityId) );

		var diplayNameIdMap = CreateMap<Entity, DisplayNameIdModel>();
		diplayNameIdMap.ForAllMembers(opt => opt.Ignore());
		diplayNameIdMap.IncludeBase<Entity, IdModel>();
		diplayNameIdMap.ForMember( target => target.DisplayName, m => m.MapFrom( source => source.DisplayName) );

		var isActiveDisplayNameIdMap = CreateMap<Entity, IsActiveDisplayNameIdModel>();
		isActiveDisplayNameIdMap.ForAllMembers(opt => opt.Ignore());
		isActiveDisplayNameIdMap.IncludeBase<Entity, DisplayNameIdModel>();
		isActiveDisplayNameIdMap.ForMember(target => target.IsActive, m => m.MapFrom(source => source.IsActive));

		var baseMap = CreateMap<Entity, BaseModel>();
		baseMap.ForAllMembers(opt => opt.Ignore());
		baseMap.IncludeBase<Entity, IsActiveDisplayNameIdModel>();
		baseMap.ForMember( target => target.Connection, m => m.MapFrom( source => source.Connection) );
		baseMap.ForMember( target => target.Type, m => m.MapFrom( source => source.Type) );

		var map = CreateMap<Entity, Model>();
		map.ForAllMembers(opt => opt.Ignore());
		map.IncludeBase<Entity, BaseModel>();
		map.ForMember( target => target.Parent, m => m.MapFrom( source => GetParentObject(source)) );

		var detailsMap = CreateMap<Entity, DetailModel>();
		detailsMap.ForAllMembers(opt => opt.Ignore());
		detailsMap.IncludeBase<Entity, Model>();
		detailsMap.ForMember( target => target.Resource, m => m.MapFrom( source => source.Resource) );
	}
}
```
I tried searching SO, here, and on Google, but with not much luck. If you have any search suggestions/combinations, I'd be happy to try those! Thanks!

###### Update 1: I also tried this pattern as well (replacing each block above) ... without success: ######
```
CreateMap<Entity, BaseModel>()
	.IncludeBase<Entity, IsActiveDisplayNameIdModel>()
	.ForMember( target => target.Connection, m => m.MapFrom( source => source.Connection) )
	.ForMember( target => target.Type, m => m.MapFrom( source => source.Type) )
	.ForAllOtherMembers(opt => opt.Ignore());
```
###### Update 2: Continued research ######
Looks like the culprit is that my "IncludeBase"s are not working properly. Not sure what other configurations to attempt.

**Comment by AlonCG** [comment]
I've seen that answer posted for any time there seems to be "explicit" mappings. I don't really think it is valid. Anyhow ... if it worked previously, then I think it should continue to work ... as I stated it worked fine in 4.1.1, does not work in 5.x. It may be a case of rolling back to 4.1.1 and being locked to that version (which is sad by acceptable) ... I really just think it is a matter of figuring out the proper configuration. And that seems to be the Issue.

**Comment by lbargaoanu** [comment]
Everything is _explicit_ here. But this is just good advice, fell free to ignore it :) I didn't say we won't help, that's why I asked for a repro.

**Comment by AlonCG** [comment]
Yes, I understand that. Wasn't criticizing your response, just remarking that every time I see that response, the OP doesn't usually accept it ... spent over 12 hours on this specific issue already. The profiles used to work, though I understand they are not ideal. Most of the mappings don't really need to be explicit, but was not sure how to do them anyway. It was just what worked. Only the "base" models really need to have properties "MapFrom" and exclude all other properties. Anyhow ... Once I can figure out how to create a gist .. I will. 

I understand how my update 1 and 2 work (or don't work) together .. since when I read it in plain English ... there is a "duh" moment. And that is ... when using "ForAllOtherMembers" the "IncludeBase" mappings seem to be ignored (which may make sense) ... though the "Inc;ludeBase" was already defined ... I would have thought they were not in the 'ForOtherMembers". I can always skip the Include base and just copy everything by hand .. but that would affect 70+ CreateMap statements.

Anyhow ... within the next few hours I will try to get a Gist ... or a standalone solution that you can run.

**Comment by AlonCG** [comment]
Update: Changing my Profile to look like
```
CreateMap<DataConnection, BaseModel>(MemberList.None)
	.IncludeBase<DataConnection, IsActiveDisplayNameIdModel>()
	.ForMember( target => target.ConnectionType, m => m.MapFrom( source => source.ConnectionType) )
	.ForMember( target => target.Description, m => m.MapFrom( source => source.Description) );
```
seems to have gotten me much further. Most likely will consider a new "mapping" strategy when we refactor our application to use DotNet core. Now ... if I could just find out how to access my "Mapper" from DI in a profile ;-) 

Just need to test a bit more ... then will most likely close this down.

**Comment by lbargaoanu** [comment]
The ForMember calls are not needed. It should work by default, they have the same name. Not sure what you mean with the mapper bit. Your mapper object should be registered as a singleton in your container.

**Comment by AlonCG** [comment]
Yeah ... that makes obvious sense. In the 4.x version, we ignored all members, so those were adding back the ones we wanted mapped. Will clean up the profiles tremendously. 

Now that I am using DI for the mapper, (nInject) I am not entirely sure how to correctly accmplish this map:
```
CreateMap<Core.Entities.Dashboards.Dashboard, DashboardCanvasesModel>()
	.ConstructUsing(source => new DashboardCanvasesModel {
		Id = source.EntityId,
		...,
		Parent = Mapper.Map<Client, ClientModel>(source.Client),
		Children = Mapper.Map<IEnumerable<Canvas>, IEnumerable<CanvasModel>>(source.Canvases)
	})
	.ForAllMembers(opt => opt.Ignore());
``` 
I know again with the opt.Ignore(). These worked well in the past, even if we did not really configure it properly. So I guess it is basically how to accomplish the above with the latest. 

I am working on trying to actually figure out how to configure the mappings properly (instead of just implementing someone's SO or blog solution) ... so if I come up with something, I'll let you know.

Thanks for all your assistance so far.

**Comment by lbargaoanu** [comment]
That looks like [constructor mapping](https://github.com/AutoMapper/AutoMapper/wiki/Construction).

**Comment by AlonCG** [comment]
Maybe?

**Comment by jbogard** [comment]
It might help to post a more full example, between ConstructUsing looking like it does the whole mapping, ForAllMembers doing Ignore...I can't tell what you actually want to happen during mapping.

**Comment by AlonCG** [comment]
Yeah I agree, just needed to try and get a working solution, sorry (everyone is waiting for this basically since I was too stubborn to roll back to the 4.x implementation). I still plan on creating a basic solution to see if what I expect the issue to be. 

So ... what you are seeing above between the "ConstructUsing" and the "ForAllMembers/IncludeBase" is basically a legacy (v1) implementation versus a v2 implementation. All the "ConstuctUsing" mappings are being migrated to the "ForAllMembers/IncludeBase" version. I am now working on ... maybe an even more "correct" v3.  Here is how I am progressing for the v3 version ... with @lorenpaulsen direction pointing and some sweat and tears ...
```
public class SetProfile : Profile
{
	public SetProfile()
	{
		CreateMap<DataSet, IdModel>(MemberList.Destination)
			.ForMember( target => target.Id, m => m.MapFrom( source => source.DataSetId ) );

		CreateMap<DataSet, DisplayNameIdModel>(MemberList.Destination)
			.IncludeBase<DataSet, IdModel>();

		CreateMap<DataSet, IsActiveDisplayNameIdModel>(MemberList.Destination)
			.IncludeBase<DataSet, DisplayNameIdModel>();

		CreateMap<DataSet, ParentedDisplayNameIdModel>(MemberList.Destination)
			.IncludeBase<DataSet, DisplayNameIdModel>()
			.ForMember( target => target.Parent, m => m.MapFrom( source => GetParentObject(source)) );

		CreateMap<DataSet, BaseModel>(MemberList.None)
			.IncludeBase<DataSet, IsActiveDisplayNameIdModel>();

		CreateMap<DataSet, Model>(MemberList.None)
			.IncludeBase<DataSet, BaseModel>()
			.ForMember( target => target.Parent, m => m.MapFrom( source => GetParentObject(source)) );

		CreateMap<DataSet, DetailModel>(MemberList.None)
			.IncludeBase<DataSet, Model>()
			.ForMember( target => target.Children, m => m.MapFrom( source => GetChildObjects(source)) );
	}

	private static DisplayNameIdModel GetParentObject(DataSet source)
	{
		var model = new DisplayNameIdModel { Id = source.DataQueryId  };
		if (source.DataQuery != null) {
			model.DisplayName = source.DataQuery.DisplayName;
		}
		return model;
	}

	private static IEnumerable<Column.BaseModel> GetChildObjects(DataSet source)
	{
		if (source.DataSetColumns == null) return null;
		var models = Mapper.Map<IEnumerable<DataSetColumn>, IEnumerable<Column.BaseModel>>(source.DataSetColumns);
		return models;
	}
}
```
This has gotten me almost to back where I was with version 4.1 (and has resulted in much cleaner profiles I might add ... thanks for that. ) 

The only thing I think I need to figure out is a better implementation for GetParentObject (though this does work somewhat) and how to implement something (a correct way to implement) for the GetChildObjects method. If I don't need to do it manually that would be great ... but I assume that initial approach of Mapping an object inside a profile was probably not a good idea to begin with? I can possibly use my IMapper constant inside the "Profile" but not sure that sounds correct.

**Comment by jbogard** [comment]
One other thing - your MapFroms should not try to access a static mapper. If you need that, go through ResolveUsing, which passes you a ResolutionContext and there you can access the Mapper.

**Comment by jbogard** [comment]
And that's because MapFrom is really about "set this mapping to be from this other member" and ResolveUsing is about "I got this, just let me give you a custom function".

**Comment by lbargaoanu** [comment]
The null checks should be added for you, but do check that you don't get exceptions at runtime. Those will be caught for you, but you want to avoid that. You shouldn't have to call Map, just map from with the right source expression. 

       MapFrom(source=>new DisplayNameIdModel { Id = source.DataQueryId,  DisplayName = source.DataQuery.DisplayName });
       MapFrom(source=>source.DataSetColumns);

**Comment by AlonCG** [comment]
@jbogard ... Regarding your last .. yeah, that is what I figured going through this process the last 24 hours-ish. That is the thingy I am trying to solve and is the last one to conquer. In my case now with 5.0.2, the StaticMapper does not work anyway in the MapFrom (may have been just luck in the 4.1 version and not really knowing how to implement what I wanted the correct way).

**Comment by lbargaoanu** [comment]
I think the first MapFrom is too complex for automatic null checks, it will just catch null exceptions, but the second should work as is. Anyway, test it :)

**Comment by AlonCG** [comment]
98% updated and working? Though I am having trouble (brain cloudy) trying to use:
```
IMappingExpression<TSource, TDestination> ConstructUsing(Func<TSource, ResolutionContext, TDestination> ctor);
```
I know how to use 
```
IMappingExpression<TSource, TDestination> ConstructUsing(Func<TSource, TDestination> ctor);
```
but have not been able to find a decent example on how to use the 1st version. I think I want to pass the ResolutionContext into my TDestination model.

**Comment by jbogard** [comment]
Correct way to call the first? So the second is:

`cfg.CreateMap<Foo, Bar>().ConstructUsing(src => new Bar());`

The first is:

`cfg.CreateMap<Foo, Bar>().ConstructUsing((src, ctxt) => new Bar { Baz = ctxt.Mapper.Map<Foo, Baz>(src) });`

**Comment by AlonCG** [comment]
Awesome thanks ... did not see this until just now (for some reason) ...

Quite possibly my initial 4.1.1 "working" solution was a hack ... and I now have a proper working configuration/mapping solution (based on your assistance). Should I still try to duplicate the initial issue in a solution, or just move on (meaning the initial issue was not really an issue)?

**Comment by lbargaoanu** [comment]
This can be closed.

**Comment by AlonCG** [comment]
Thanks guys for putting up with me ... keep up the good work and all the best!
